### PR TITLE
addl unit tests for PR #3970

### PR DESCRIPTION
**Pull request by rwgk**
## Description
Contribution to / preparation for PR #3970. The `inc_refs` line in test_pytypes.py will need to be adjusted, which will effectively document the behavior change.

Full test coverage for all newly added functions, sensitive to the change from copy to move behavior.

I went through top-to-bottom in this copy of pytypes.h, undoing the changes under PR #3970 one-by-one, and verifying that the new test fails.

The newly introduced `PYBIND11_HANDLE_REF_DEBUG` might be useful for unit-testing other move-instead-of-copy optimizations. The main goal is to ensure that such optimizations do not accidentally get undone.

## Suggested changelog entry:

```rst

```

### CODE REVIEW

File: include/pybind11/pytypes.h [hunk 1]
```diff
@@ -180,6 +180,10 @@ class object_api : public pyobject_tag {
 
 PYBIND11_NAMESPACE_END(detail)
 
+#if !defined(PYBIND11_HANDLE_REF_DEBUG) && !defined(NDEBUG)
+#    define PYBIND11_HANDLE_REF_DEBUG
```
Comment (leslinice) [inline include/pybind11/pytypes.h:184]: LT

File: include/pybind11/pytypes.h [hunk 2]
```diff
@@ -244,6 +251,23 @@ class handle : public detail::object_api<handle> {
...
+private:
+#    if defined(WITH_THREAD)
+#        error LOOOK_WITH_THREAD
+#    else
+#        error LOOOK_NOT_WITH_THREAD
+#    endif
+    std::size_t static inc_ref_counter(std::size_t add) {
+        thread_local std::size_t counter = 0;
+        counter += add;
+        return counter;
```
Comment (leslinice) [inline include/pybind11/pytypes.h:265]: 10000

**Review by leslinice — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]
@rwgk Looks good to me.

**Comment by rwgk** [comment]
> @rwgk Looks good to me.

Great, thanks!
Unless you started on it already, I could strip out the PR #3970 changes from here and merge.
Then you could just rebase, adjust the expected `inc_refs` line, and merge your PR.

**Pull request merged**